File: src/stores/skills.ts
import { create } from "zustand";
import { invoke } from "@/lib/tauri";

export interface Skill {
  id: string;
  name: string;
  description: string | null;
  body_bytes: number;
  path: string;
}

interface SkillInput {
  id: string;
  name: string;
  description?: string | null;
  body?: string;
}

interface SkillsState {
  skills: Skill[];
  loading: boolean;
  list: () => Promise<void>;
  create: (input: SkillInput) => Promise<void>;
  update: (input: SkillInput) => Promise<void>;
  remove: (id: string) => Promise<void>;
  setEnabled: (id: string, enabled: boolean) => Promise<void>;
  readSource: (id: string) => Promise<string>;
}

/**
 * User-authored skills. Mirrors `crate::skills` on the Rust side: each skill
 * is a folder under `~/.zero/skills/<id>/` containing `SKILL.md`. Enabled
 * ids are stored in `settings.skills_enabled` (see `useSettingsStore`) so
 * the chat runner can find them every turn.
 */
export const useSkillsStore = create<SkillsState>((set) => ({
  skills: [],
  loading: false,
  list: async () => {
    set({ loading: true });
    try {
      const skills = (await invoke<Skill[]>("skills_list")) ?? [];
      set({ skills, loading: false });
    } catch (e) {
      console.error("skills_list failed", e);
      set({ loading: false });
    }
  },
  create: async (input) => {
    await invoke("skills_create", { input });
    const skills = (await invoke<Skill[]>("skills_list")) ?? [];
    set({ skills });
  },
  update: async (input) => {
    await invoke("skills_update", { input });
    const skills = (await invoke<Skill[]>("skills_list")) ?? [];
    set({ skills });
  },
  remove: async (id) => {
    await invoke("skills_delete", { id });
    set((s) => ({ skills: s.skills.filter((sk) => sk.id !== id) }));
  },
  setEnabled: async (id, enabled) => {
    await invoke("skills_set_enabled", { id, enabled });
  },
  readSource: async (id) => {
    return (await invoke<string>("skills_read_source", { id })) ?? "";
  },
}));

File: src-tauri/src/commands/skills.rs
//! Tauri commands for user-authored skills.

use crate::error::IpcResult;
use crate::settings::Settings;
use crate::skills::{self, Skill};

#[tauri::command]
pub async fn skills_list() -> IpcResult<Vec<Skill>> {
    skills::list().await.map_err(|e| e.to_string().into())
}

#[derive(Debug, serde::Deserialize)]
pub struct SkillInput {
    pub id: String,
    pub name: String,
    #[serde(default)]
    pub description: Option<String>,
    #[serde(default)]
    pub body: String,
}

#[tauri::command]
pub async fn skills_create(input: SkillInput) -> IpcResult<Skill> {
    skills::create(
        &input.id,
        &input.name,
        input.description.as_deref(),
        &input.body,
    )
    .await
    .map_err(|e| e.to_string().into())
}

#[tauri::command]
pub async fn skills_update(input: SkillInput) -> IpcResult<Skill> {
    skills::update(
        &input.id,
        &input.name,
        input.description.as_deref(),
        &input.body,
    )
    .await
    .map_err(|e| e.to_string().into())
}

#[tauri::command]
pub async fn skills_delete(id: String) -> IpcResult<()> {
    skills::delete(&id).await.map_err(|e| e.to_string().into())
}

/// Toggle a skill's enabled state in `settings.skills_enabled`. The runner
/// reads this list every turn so changes take effect on the next message.
#[tauri::command]
pub async fn skills_set_enabled(id: String, enabled: bool) -> IpcResult<()> {
    let mut s = Settings::load().await.map_err(|e| e.to_string())?;
    let has = s.skills_enabled.iter().any(|x| x == &id);
    if enabled && !has {
        s.skills_enabled.push(id);
    } else if !enabled && has {
        s.skills_enabled.retain(|x| x != &id);
    } else {
        return Ok(());
    }
    s.save().await.map_err(|e| e.to_string().into())
}

/// Return the raw `SKILL.md` text. Used by the editor in the Skills page.
#[tauri::command]
pub async fn skills_read_source(id: String) -> IpcResult<String> {
    let s = skills::load_with_body(&id)
        .await
        .map_err(|e| e.to_string())?;
    Ok(tokio::fs::read_to_string(&s.meta.path)
        .await
        .map_err(|e| e.to_string())?)
}
